cmd/kubeadm/app/phases/kubelet/flags_windows.go

```diff
@@ -0,0 +1,64 @@
+// +build windows
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"strings"
+
+	"k8s.io/klog"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+// buildKubeletArgMap takes a kubeletFlagsOpts object and builds based on that a string-string map with flags
+// that should be given to the local kubelet daemon.
+func buildKubeletArgMap(opts kubeletFlagsOpts) map[string]string {
+	kubeletFlags := map[string]string{}
+
+	if opts.nodeRegOpts.CRISocket == constants.DefaultDockerCRISocket {
+		// These flags should only be set when running docker
+		kubeletFlags["network-plugin"] = "cni"
+		if opts.pauseImage != "" {
+			kubeletFlags["pod-infra-container-image"] = opts.pauseImage
+		}
+	} else {
+		kubeletFlags["container-runtime"] = "remote"
+		kubeletFlags["container-runtime-endpoint"] = opts.nodeRegOpts.CRISocket
+	}
+
+	if opts.registerTaintsUsingFlags && opts.nodeRegOpts.Taints != nil && len(opts.nodeRegOpts.Taints) > 0 {
+		taintStrs := []string{}
+		for _, taint := range opts.nodeRegOpts.Taints {
+			taintStrs = append(taintStrs, taint.ToString())
+		}
+
+		kubeletFlags["register-with-taints"] = strings.Join(taintStrs, ",")
+	}
+
+	// Pass the "--hostname-override" flag to the kubelet only if it's different from the hostname
+	nodeName, hostname, err := GetNodeNameAndHostname(opts.nodeRegOpts)
+	if err != nil {
+		klog.Warning(err)
+	}
+	if nodeName != hostname {
+		klog.V(1).Infof("setting kubelet hostname-override to %q", nodeName)
+		kubeletFlags["hostname-override"] = nodeName
+	}
+
+	return kubeletFlags
```

Comment:
overall i'm in favor of this change in particular.

but it makes sense to do the following:
- have one big buildKubeletArgMapCommon() function with a common setup:
  - hostname override setup
  - ipv6dualstack FG setup?
  - taints setup
- have abstracted buildKubeletArgMap() function on Windows that has:
  - CRI setup
- have abstracted buildKubeletArgMap() function on Linux that has:
  - CRI setup
  - resolve conf setup

Reply:
`resolv.conf` and `--pod-infra-container-image` can be counted as CRI setup (more accurately - Docker backend setup).
Ideally we should have these isolated somewhere even on Linux, as they are unlikely to get into the Kubelet component config. If the Docker shim moves, they are going to move with it.
Then the only thing left to do is the CRI. This, however, still does not imply having separate `buildKubeletArgMap` versions for Windows and Linux.

Reply:
we had a discussion with @rosti about this during the kubeadm office hours and my proposal above to split the functions seems valid.

later on we want to move flags that are already in the v1beta1 KubeletConfiguration to the config.

Reply:
The Windows version of `buildKubeletArgMap` is now a subset of the Linux one. How about we move it into a common space and name it `buildKubeletArgMapCommon`. Then have the Linux version of `buildKubeletArgMap` call it and add the rest of the flags. The Windows version of `buildKubeletArgMap` would be just a one liner (call `buildKubeletArgMapCommon`).

@gab-satchi @neolit123 WDYT?

Reply:
```
	if opts.nodeRegOpts.CRISocket == constants.DefaultDockerCRISocket {
		// These flags should only be set when running docker
		kubeletFlags["network-plugin"] = "cni"
		if opts.pauseImage != "" {
			kubeletFlags["pod-infra-container-image"] = opts.pauseImage
		}
	} else {
		kubeletFlags["container-runtime"] = "remote"
		kubeletFlags["container-runtime-endpoint"] = opts.nodeRegOpts.CRISocket
	}
```

can be moved to common.
then Linux has to have the explicit:

```
	if opts.nodeRegOpts.CRISocket == constants.DefaultDockerCRISocket {
		driver, err := kubeadmutil.GetCgroupDriverDocker(opts.execer)
		if err != nil {
			klog.Warningf("cannot automatically assign a '--cgroup-driver' value when starting the Kubelet: %v\n", err)
		} else {
			kubeletFlags["cgroup-driver"] = driver
		}
	}
```

and Windows can only call buildKubeletArgMapCommon()